### ISSUE TITLE
Should PackageMetadata.get be supported?

### ISSUE THREAD

**Issue by duckinator** [python/importlib_metadata#384]
This isn't a feature request. I'm looking for a quick clarification, so other folks who encounter python/typeshed#7767 have a definitive answer.

---

In the wild, I've seen things like this:

```
from importlib.metadata import metadata
print(metadata("bork").get("Version"))
```

Both the [stdlib documentation](https://docs.python.org/3/library/importlib.metadata.html) and the [readthedocs documentation](https://importlib-metadata.readthedocs.io/en/latest/) never mention `.get()`, and `PackageMetadata` does not define it. Due to that, using it and running mypy returns `error: "PackageMetadata" has no attribute "get"`. I suspect I made a bad assumption when I used it, but if so it's one I've seen other folks make &mdash; so it seems worthwhile to get an explicit answer on whether leaving it out is deliberate.

**So, my question is this:**

Is `.get()` an implementation detail that folks latched on to and should migrate away from? Or should it be added to `PackageMetadata`?

**Comment by rra** [comment]
The specific use case I have for using `.get()` is that sometimes I'm retrieving package metadata that may or may not exist (in support libraries that use package metadata for various purposes, for example). Without `.get()`, this would require a `try/catch` block around a hash dereference. That can be awkward, particularly when retrieving multiple bits of metadata at the same time and leaving the ones that aren't set as `None`.

**Comment by jaraco** [comment]
tl;dr - yes, it's an implementation detail.

This is another case (similar to #371), where the duck-typing approach has left ambiguity. This project has provided a viable implementation but not a rigorously specified one. In more recent releases, `PackageMetadata` was introduced to be explicit about the supported interface. That answers the question in the title. It's an implementation detail that was not expected to be used.

That said, perhaps PackageMetadata should supply `.get()`. For the standard providers, because that method already exists in this implementation. The challenge is that if there are other providers that don't provide .get and the protocol is changed, that implementation won't be in compliance with the protocol. So we'll want a compelling case why `.get` is valuable.

What about using the `.json` accessor? It specifies a Dict object, which does have `.get`. Could that be suitable for your needs?

Or alternately, does `PackageMetadata` provide enough of `Mapping` to be able to be cast to a `Dict`? It seems it does, as it supplies all of the [abstract methods](https://docs.python.org/3/library/collections.abc.html#collections-abstract-base-classes) of `Mapping`. That may mean that there's a straightforward way to create an adapter that would provide .get over a PackageMetadata.

**Comment by duckinator** [comment]
@jaraco thank you for the clarification, I appreciate it! My usecase was resolved by switching from `.get(key)` to `[key]`, and I was mainly just trying to get clarification for python/typeshed#7767. So my input on `.get()` itself isn't really relevant. I'll leave the discussion of that to the people it actually affects. :)

**Comment by jaraco** [comment]
> Or alternately, does `PackageMetadata` provide enough of `Mapping` to be able to be cast to a `Dict`? It seems it does, as it supplies all of the [abstract methods](https://docs.python.org/3/library/collections.abc.html#collections-abstract-base-classes) of `Mapping`. That may mean that there's a straightforward way to create an adapter that would provide .get over a PackageMetadata.

Let me illustrate what I was suggesting here with an example:

```python
description = dict(metadata("bork")).get("Description")
```

Granted, that approach is a bit convoluted and not performant (creates an entire copy of the metadata to extract one value).

There seems to be a reasonable expectation that `PackageMetadata` should provide `.get`, especially after `.__getitem__` starts raising `KeyError`, so let's proceed in that direction and see what happens.
